![](https://www.dii.uchile.cl/wp-content/uploads/2021/06/Magi%CC%81ster-en-Ciencia-de-Datos.png)

**MDS7202: Laboratorio de Programación Científica para Ciencia de Datos**

### 👨‍🏫👩‍🏫 Cuerpo Docente:

- Profesor: Pablo Badilla, Diego Cortez
- Auxiliar: Melanie Peña Torres, Valentina Rojas Osorio
- Ayudante: Javiera Arévalo, Tamara Carrasco, Ignacio Reyes

### 👨‍💻👩‍💻 Estudiantes:
- Estudiante n°1: Isadora Madrid Espinoza

_Por favor, lean detalladamente las instrucciones de la tarea antes de empezar a escribir._

---

## Reglas

El proyecto debe realizarse al final del curso y su objetivo es poner a prueba el conocimiento de todas las unidades del curso. Este consta de 3 partes: **Informe de Análisis** (30%), **Construcción del Modelo** (40%) y **Despliegue del Modelo** (30%). Este notebook corresponde a la **Parte 1 - Informe de Análisis**.

### **Requisitos del proyecto**
- **Grupos**: Formar equipos de **2 personas**. 
- **Plazo de entrega**: Hasta el Viernes 10 de Julio
- **Consultas**: Cualquier duda fuera del horario de clases debe ser planteada en el foro correspondiente. Los mensajes enviados al equipo docente serán respondidos únicamente por este medio. Por favor, revisen las respuestas anteriores en el foro antes de realizar nuevas consultas.
- **Plagio**: La copia o reutilización no autorizada de trabajos de otros grupos está **estrictamente prohibida**. El incumplimiento de esta norma implicará la anulación inmediata del proyecto y una posible sanción académica.
- **Material permitido**: Pueden usar cualquier material del curso, ya sea notas, lecturas, códigos, o referencias proporcionadas por los docentes, que consideren útil para el desarrollo del proyecto.

---

## 📊 Distribución de Puntaje

Esta parte se evalúa en **escala de 1.0 a 7.0**, donde la nota se calcula como:

$$\text{Nota} = 1.0 + \text{puntaje obtenido}$$

El puntaje total repartible es de **6.0 puntos**, distribuidos así:

| Sección | Detalle | Puntaje |
| ------- | ------- | ------: |
| **Sección 1 - Entendimiento del Problema** | Preguntas conceptuales (P1–P7) | 2.0 |
| **Sección 2 · Subsección 1 - Datos Tabulares** | Código (1.0) + Preguntas (1.6) | 2.6 |
| **Sección 2 · Subsección 2 - Embeddings** | Código (0.6) + Preguntas (0.8) | 1.4 |
| **Total** | | **6.0** |

Cada pregunta y tarea indica su puntaje entre paréntesis, p. ej. _(0.3 pts)_.


Esta nota corresponde a la nota de la **Parte 1** que equivale a un **30%** de la nota del proyecto final. Para obtener la nota final, las notas individuales de cada parte se ponderan de la siguiente manera.

| Parte              | Ponderación |
|--------------------|------------:|
| **1 — Análisis**   | **30%**     |
| 2 — Modelo         | 40%         |
| 3 — Despliegue     | 30%         |

---

## 📖 Introducción 

<p align="center">
    <img src="./chaucherapp.jpg" alt="Alt Text" width="700">
</p>

La app chilena **ChaucherApp** ha emergido recientemente como una de las nuevas apps *FinTech* que apuesta por dominar en un competitivo mercado. Esta app ofrece una billetera digital que permite hacer transferencias, tarjetas digitales que permiten comprar, y servicios financieros como préstamos, seguros e inversiones.

Sin duda es sorprendente que **ChaucherApp** sea uno de los líderes de la industria fintech, siendo una empresa tan joven. En las palabras del CEO de la empresa, el Sr. Querty _'El gran éxito de ChaucherApp se debe a que entendimos la importancia de escuchar y responder eficientemente al cliente'_. En efecto, **ChaucherApp** fue diseñada desde el comienzo con una robusta arquitectura digital relacionada al **Soporte al cliente**.

Durante años, **ChaucherApp** ha logrado dar respuesta satisfactoria a cientos de tickets de soporte al cliente al mes, lo que le ha permitido inspirar confianza y prestigio. Por supuesto, esto también ha generado una gran cantidad de datos que pueden ser útiles para las siguientes etapas de crecimiento de la empresa. Y es que, a medida que el mercado se hace más competitivo, se hace más necesario aumentar la eficiencia de los procesos de respuesta de tickets, por lo que el Sr. Querty ha optado por **automatizar el proceso** incorporando inteligencia artificial.

Luego de intensa planificación ejecutiva y jornadas de analítica, han decidido que el primer paso para automatizar será el de desarrollar un modelo de **priorización de tickets** de soporte al cliente utilizando **modelos de lenguaje** y **aprendizaje de máquinas**. De esta manera podrán optimizar recursos humanos mientras se desarrollan otras herramientas.

Para lograr esta tarea, te han encargado analizar datos históricos de tickets de soporte al cliente que ya han sido manualmente priorizados, presentes en el archivo `data/tickets.parquet`. En él, encontrarás tickets con el contenido del ticket, categoría, canal, fecha y priorización, que puede ser `Baja`, `Media`, `Alta` o `Crítica`. Tu objetivo es utilizar esa data y los medios, algoritmos y estrategias que estimes convenientes para construir un **modelo clasificador** que sea capaz de predecir esta categoría en base al texto y atributos del ticket.

Con esta información, tus tareas a realizar son primero **analizar y entender los datos**, luego **construir un modelo de inferencia**, y finalmente **implementar ese modelo** para que el equipo de ejecutivos de ChaucherApp pueda utilizarlo.

---

# Sección 1 - Entendimiento del Problema

_(Total: 2.0 pts)_

El objetivo de esta sección es entender cuál será el problema que tendrás que resolver, incluyendo tanto la definición misma del problema (i.e., si es clasificación binaria, clasificación multiclase, regresión, clustering, otros...) como también las métricas, tanto de negocio como de evaluación, que usarás para medir el éxito de tu solución.

Para esto, primero responde las siguientes preguntas:


1. ¿Qué tipo de problema vamos a resolver? ¿Clasificación? ¿Regresión? ¿Clustering? ¿Reducción de dimensionalidad? ¿Ranking? ¿Recomendación? ¿Por qué? (utiliza la definición de cada uno de estos tipos de problemas más el enunciado para fundamentar). _(0.3 pts)_
2. ¿Por qué el negocio te solicitaría resolver este problema? ¿Por qué es importante? _(0.2 pts)_
3. ¿Por qué el negocio solicita generar un modelo predictivo para resolver este problema? ¿Por qué no lo pueden hacer, por ejemplo, por medio de código más palabras claves? _(0.2 pts)_
4. ¿Cuáles serían las métricas de evaluación relevantes para el negocio que buscarás optimizar? Aquí tienes la libertad de fundamentar lo que se te ocurra en base a tu entendimiento del negocio. Nombra al menos 3 y justifica la elección. _(0.3 pts)_
5. ¿Cuáles serían las métricas de evaluación de ML importantes para este problema? Para responder esto, primero da la interpretación de qué significa cada métrica (R^2, RMSE, MSE, etc.) / (Accuracy, Precision, Recall, F1) respecto a un problema binario Correcto / Incorrecto. Importante: Debes responder con la interpretación de qué significa cada una de estas métricas respecto al problema binario. _(0.5 pts)_
6. Luego, a partir de las interpretaciones anteriores de métricas más el conocimiento del negocio, elige cuál métrica es la más apropiada para cada clase en base a las prioridades del negocio. _(0.3 pts)_
7. Por último, ¿qué es la agregación macro? ¿Y la weighted? Y, ¿qué tipo de agregación utilizarías para promediar estas métricas de evaluación? _(0.2 pts)_

> 1. El problema corresponde a un problema de clasificación supervisada multiclase. Es supervisada porque se cuenta con datos históricos de tickets que ya fueron priorizados manualmente, por lo tanto existe una variable objetivo conocida. Es multiclase porque el modelo debe asignar cada ticket a una de cuatro categorías posibles de prioridad: Baja, Media, Alta o Crítica.
No corresponde a un problema de regresión, ya que no se busca predecir un valor numérico continuo, como un monto o un tiempo estimado de resolución. Tampoco corresponde a clustering, porque no se busca descubrir grupos desconocidos dentro de los datos, sino aprender a partir de clases previamente definidas. Podría existir un enfoque de ranking para ordenar tickets según urgencia, pero el enunciado nos pide explícitamente predecir una categoría de prioridad, por lo que el enfoque más adecuado es clasificación.
El problema se relaciona con procesamiento de lenguaje natural, ya que una parte importante de la información viene del texto escrito por los clientes en los tickets. La literatura sobre clasificación automática de tickets de soporte señala que automatizar este proceso puede mejorar el desempeño del área de atención al cliente y reducir tiempos de resolución. (señalo la pagína que encontre con información con respecto a este tema arvix.org) .

> 2. El negocio solicita resolver este problema porque ChaucherApp recibe una gran cantidad de tickets de soporte y necesita asignar sus recursos de manera eficiente. No todos los tickets tienen la misma urgencia ni el mismo impacto, por ejemplo, un problema de consulta general no debería recibir la misma prioridad que un ticket asociado a una falla de pago, bloqueo de cuenta o fraude.
La correcta priorización permite que los tickets más urgentes sean atendidos antes, reduciendo riesgos operacionales, pérdidas de confianza y deterioro en la experiencia del cliente. Sobre todo esto es importante porque una mala atención al cliente puede afectar directamente la reputación de la empresa, la retención de usuarios y la percepción de seguridad del servicio.

> 3. El negocio solicita un modelo predictivo porque los tickets de soporte están escritos en lenguaje natural, por lo que pueden expresar el mismo problema de muchas formas distintas. Por ejemplo, un usuario podría escribir “no puedo pagar”, “mi tarjeta no funciona”, “la app rechaza la compra” o “me aparece error al comprar”, y todas esas frases podrían referirse a problemas similares, aunque no usen exactamente las mismas palabras.
Un sistema basado solo en palabras clave sería rígido, difícil de mantener y propenso a errores. Tendría que anticipar manualmente todas las formas posibles en que los clientes describen un problema. 
En cambio, un modelo de ML puede aprender patrones desde los datos históricos, considerando combinaciones de palabras, contexto y atributos adicionales. Esto permite una clasificación más flexible y adaptable. 

> 4. Las métricas relevantes que yo buscaría optimizar serían:
 >- Tiempo de primera respuesta por prioridad: mide cuánto tarda el equipo en dar la primera respuesta a un ticket desde que fue ingresado. Es especialmente relevante para tickets de prioridad Alta o Crítica, ya que estos deberían ser atendidos rápidamente. Sobre todo porque las personas no son pacientes entonces reducir el tiempo de respuesta permitira que tengan una mejor percepción de la emrpesa.
> - Tiempo promedio o mediano de resolución: mide cuánto tarda la empresa en resolver completamente los tickets. Si el modelo prioriza correctamente, debería ayudar a reducir el tiempo de resolución de los tickets más urgentes.
> - Tiempo de resolución de ticket por clase de prioridad: midiendo qué porcentaje de tickets se resuelve dentro del tiempo comprometido para cada nivel de prioridad. Esto es importante porque no todas las prioridades deberían tener el mismo plazo de atención. 
> - Satisfacción del cliente: permite evaluar si la automatización realmente mejora la experiencia del usuario. Un modelo podría ser técnicamente bueno, pero si genera malas asignaciones o retrasa tickets urgentes, la satisfacción podría disminuir.

> 5. Como este es un problema de clasificación multiclase, las métricas más importantes son accuracy, precision, recall y F1-score. Para interpretarlas de manera simple, se puede pensar primero en un problema binario, una predicción puede estar Correcta o Incorrecta. Luego, en el caso multiclase, estas métricas se calculan para cada clase usando una lógica uno contra los demas, por ejemplo, “Crítica” versus “No Crítica”.
> - Accuracy: mide la proporción total de tickets correctamente clasificados. Es decir, de todos los tickets evaluados, cuántos fueron asignados a la prioridad correcta. Scikit-learn define accuracy como la fracción de predicciones correctas sobre el total de muestras.
> - Precision: mide, para una clase específica, qué proporción de los tickets que el modelo predijo como esa clase realmente pertenecían a ella. Por ejemplo, si el modelo predice 100 tickets como Críticos y 80 efectivamente eran Críticos, la precision de la clase Crítica sería 80%. Esta métrica ayuda a controlar los falsos positivos.
> - Recall: mide, para una clase específica, qué proporción de los tickets que realmente pertenecían a esa clase fueron correctamente detectados por el modelo. Por ejemplo, si existían 100 tickets realmente Críticos y el modelo detectó 85, el recall de la clase Crítica sería 85%. Esta métrica es muy importante cuando el costo de no detectar una clase es alto.
> - F1-score: combina precision y recall en una sola métrica mediante un promedio armónico. Es útil cuando se busca equilibrio entre no sobreclasificar una clase y no dejar casos importantes sin detectar. 
> - MSE: mide el promedio de los errores al cuadrado entre el valor real y el valor predicho. Se utiliza cuando se quiere predecir un número, por ejemplo, el tiempo de resolución de un ticket en horas. Mientras más bajo sea el MSE, mejor es el modelo. 
> - RMSE: corresponde a la raíz cuadrada del MSE. Tiene la ventaja de quedar expresado en la misma unidad de la variable que se predice. Por ejemplo, si se estuviera prediciendo el tiempo de resolución en horas, el RMSE también estaría expresado en horas.
> - R²: mide qué proporción de la variabilidad de una variable numérica es explicada por el modelo. Un R² cercano a 1 indica que el modelo explica bien los datos, mientras que un valor cercano a 0 indica bajo poder explicativo. Esta métrica se utiliza en regresión, por ejemplo, para evaluar qué tan bien se predice el tiempo de atención de un ticket.
> En este problema, métricas de regresión como MSE, RMSE o R² no son las más adecuadas, porque la variable objetivo no es numérica continua, sino una etiqueta categórica de prioridad.

> 6. Dado que las clases representan distintos niveles de urgencia, no todas deberían evaluarse con el mismo criterio principal:
> - Para la clase crítica la metrica más importante es Recall, pues lo más importante es no dejar pasar tickets realmente críticos. Un falso negativo en esta clase puede significar que un caso urgente sea tratado como menos importante.
> - Para la clase alta la metrica más importante es Recall o F1-score, pues también es importante detectar la mayor cantidad posible de tickets urgentes, aunque manteniendo cierto equilibrio para no saturar al equipo con demasiados falsos positivos.
> - Para la clase media la metrica más importante es F1-score, pues en esta clase conviene equilibrar precision y recall, ya que no es tan crítica como Alta o Crítica, pero igualmente debe ser clasificada correctamente para ordenar la carga operativa.
> - Para la clase baja la metrica más importante es Precision, pues es importante que los tickets clasificados como Baja realmente sean de baja prioridad. Si un ticket urgente se clasifica erróneamente como Baja, podría quedar postergado injustificadamente.

> 7. La agregación macro calcula la métrica por separado para cada clase y luego obtiene un promedio simple, dando el mismo peso a todas las clases. Esto es útil cuando se quiere evaluar si el modelo funciona bien en todas las categorías, incluso en aquellas con menos datos.
La agregación weighted también calcula la métrica por clase, pero luego promedia ponderando por la cantidad de ejemplos reales de cada clase. Esto refleja mejor el desempeño global cuando hay clases más frecuentes, pero puede ocultar malos resultados en clases minoritarias.
Para este caso, utilizaría macro F1-score como métrica principal, porque en soporte al cliente las clases minoritarias pueden ser justamente las más importantes, como los tickets Críticos. Si solo se usara weighted F1-score, el desempeño podría verse artificialmente alto si el modelo clasifica bien muchos tickets de Baja o Media, pero falla en los Críticos.
De todas maneras, también usaría weighted F1-score como métrica complementaria, porque permite observar el rendimiento general del modelo considerando la distribución real de tickets. 

---

# Sección 2 - Análisis de Datos

## Subsección 1 - Datos Tabulares

_(Total: 2.6 pts — código 1.0 + preguntas 1.6)_


La idea de esta segunda sección es que leas las fuentes de datos y las explores.


Partiremos primero con `tickets.parquet`. Utiliza la función de I/O asociada a este tipo de archivo para cargar los datos.

In [41]:
import pandas as pd
import plotly.express as px
import plotly.graph_objects as go
from plotly.subplots import make_subplots
from pathlib import Path

# Posibles rutas, dependiendo desde dónde esté corriendo el notebook
posibles_rutas = [
    Path("data/tickets.parquet"),
    Path("Proyecto_final/data/tickets.parquet"),
]

ruta_tickets = None

for ruta in posibles_rutas:
    if ruta.exists():
        ruta_tickets = ruta
        break

if ruta_tickets is None:
    raise FileNotFoundError("No se encontró el archivo tickets.parquet. Revisa la ruta del archivo.")

# Cargar datos
df_tickets = pd.read_parquet(ruta_tickets)

# Mostrar primeras filas
df_tickets.head()



,Id_Ticket,Asunto_Ticket,Contenido_Ticket,N_Caracteres_Ticket,Canal_Ticket,Categoría_Problema,Usuario-Tipo_de_Cuenta,Usuario-Antiguedad_Cuenta_Dias,Fecha_Envío,Tiempo_de_Resolución_Horas,Nivel_Prioridad
0,TCK-19267,Error al iniciar sesión en nuevo celular por d...,hola oye lo que pasa es que me cambie de celu ...,531,Whatsapp,Cuenta,Premium,159,2024-01-02,14,Baja
1,TCK-19268,Error al actualizar correo electrónico: bloque...,"Hola, buenas tardes. Escribo porque estoy trat...",784,Whatsapp,Otro,Premium,192,2024-01-02,25,Media
2,TCK-19269,Cobro desconocido en mi tarjeta digital,"Hola buenas, escribo porque me aparecio un des...",600,Whatsapp,Fraude,Free,314,2024-01-02,10,Baja
3,TCK-19270,"Cuenta bloqueada sin aviso, necesito ayuda urg...","Hola, necesito que me solucionen esto luego po...",659,Whatsapp,Cuenta,Free,228,2024-01-02,2,Media
4,TCK-19271,Me descontaron plata del cajero y no salió el ...,Hola buenos dias escribo porque ya estoy aburr...,853,Whatsapp,Otro,Free,72,2024-01-03,24,Media


Ahora, analiza los datos de entrada usando las herramientas solicitadas _(código: 1.0 pts en total)_:


1. Utiliza `.info()` y `.describe()` para obtener una visualización rápida del dataframe. _(0.1 pts)_
2. Utiliza `histogram` de plotly express para explorar la distribución de la variable objetivo, y `value_counts` más operaciones entre series para calcular la cantidad y porcentaje de sus categorías. _(0.2 pts)_
3. Utiliza `histogram` más [subplots](https://plotly.com/python/subplots/) para observar la distribución de características numéricas y categóricas. Hint: puedes acceder con `fig.data[0]` a la traza de un histograma generado por `histogram`. _(0.2 pts)_
4. Utiliza `corr` + `imshow` para analizar la correlación entre variables numéricas. _(0.1 pts)_
5. Genera bins con `qcut` para `N_Caracteres_Ticket`, `Tiempo_de_Resolución_Horas` y `Usuario-Antiguedad_Cuenta_Dias`, con los valores `Bajo`, `Medio`, `Alto`. Luego, utiliza `crosstab` (normalizando usando `normalize` por filas) + `imshow` para analizar las contingencias entre `Categoría_Problema` y `Nivel_Prioridad`, `Canal_Ticket` y `Nivel_Prioridad`, `Usuario-Tipo_de_Cuenta` y `Nivel_Prioridad`, y los bins de `N_Caracteres_Ticket` y `Nivel_Prioridad`, `Tiempo_de_Resolución_Horas` y `Nivel_Prioridad`, y `Usuario-Antiguedad_Cuenta_Dias` y `Nivel_Prioridad`. Grafica en subplots. _(0.2 pts)_
6. Por último, agrega la cantidad de reclamos por mes y `Nivel_Prioridad` (crea una nueva columna llamada `Fecha_Envío_Mes` usando el módulo de datetimes de pandas) para obtener un conteo de cantidad de reclamos según mes-nivel de prioridad, y utiliza `line` para generar un gráfico de línea en donde cada `Nivel_Prioridad` tenga su propia serie. _(0.2 pts)_

In [42]:
############## 1 ###############
df_tickets.info()
df_tickets.describe()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 2000 entries, 0 to 1999
Data columns (total 11 columns):
 #   Column                          Non-Null Count  Dtype 
---  ------                          --------------  ----- 
 0   Id_Ticket                       2000 non-null   object
 1   Asunto_Ticket                   2000 non-null   object
 2   Contenido_Ticket                2000 non-null   object
 3   N_Caracteres_Ticket             2000 non-null   int64 
 4   Canal_Ticket                    2000 non-null   object
 5   Categoría_Problema              2000 non-null   object
 6   Usuario-Tipo_de_Cuenta          2000 non-null   object
 7   Usuario-Antiguedad_Cuenta_Dias  2000 non-null   int64 
 8   Fecha_Envío                     2000 non-null   object
 9   Tiempo_de_Resolución_Horas      2000 non-null   int64 
 10  Nivel_Prioridad                 2000 non-null   object
dtypes: int64(3), object(8)
memory usage: 172.0+ KB


,N_Caracteres_Ticket,Usuario-Antiguedad_Cuenta_Dias,Tiempo_de_Resolución_Horas
count,2000.000000,2000.000000,2000.000000
mean,717.699000,121.743000,18.718000
std,209.355096,142.339366,18.841788
min,251.000000,4.000000,1.000000
25%,564.000000,42.000000,5.000000
50%,697.000000,78.000000,13.000000
75%,850.250000,143.000000,26.000000
max,1831.000000,1200.000000,183.000000


In [43]:
################ 2 ###############
variable_objetivo = "Nivel_Prioridad"

# Cantidad por prioridad
conteo_prioridad = df_tickets[variable_objetivo].value_counts()
conteo_prioridad

#Porcentaje por prioridad
porcentaje_prioridad = df_tickets[variable_objetivo].value_counts(normalize=True) * 100
porcentaje_prioridad

# Tabla resumen
tabla_prioridad = pd.DataFrame({
    "Cantidad": conteo_prioridad,
    "Porcentaje": porcentaje_prioridad.round(2)
})

tabla_prioridad

#Histograma con Plotly Express
fig = px.histogram(
    df_tickets,
    x="Nivel_Prioridad",
    title="Distribución de la variable objetivo: Nivel de Prioridad",
    category_orders={
        "Nivel_Prioridad": ["Baja", "Media", "Alta", "Crítica"]
    }
)

fig.show()

In [ ]:
############ 3 ##########
from plotly.subplots import make_subplots
import plotly.express as px

variables_histogramas = [
    "N_Caracteres_Ticket",
    "Tiempo_de_Resolución_Horas",
    "Usuario-Antiguedad_Cuenta_Dias",
    "Canal_Ticket",
    "Categoría_Problema",
    "Usuario-Tipo_de_Cuenta"
]

# Verificar si todas las columnas existen
columnas_faltantes = [col for col in variables_histogramas if col not in df_tickets.columns]

if columnas_faltantes:
    print("Columnas que no existen en el dataframe:")
    print(columnas_faltantes)
else:
    print("Todas las columnas existen correctamente.")


fig = make_subplots(
    rows=3,
    cols=2,
    subplot_titles=variables_histogramas
)

for i, col in enumerate(variables_histogramas):
    hist = px.histogram(
        df_tickets,
        x=col
    )
    
    fila = i // 2 + 1
    columna = i % 2 + 1
    
    fig.add_trace(hist.data[0], row=fila, col=columna)

fig.update_layout(
    title="Distribución de características numéricas y categóricas",
    height=900,
    showlegend=False
)

fig.show()

Todas las columnas existen correctamente.


In [29]:
######### 4 ##########
columnas_numericas = df_tickets.select_dtypes(include=["int64", "float64"]).columns

matriz_corr = df_tickets[columnas_numericas].corr()

matriz_corr

fig = px.imshow(
    matriz_corr,
    text_auto=True,
    title="Matriz de correlación entre variables numéricas"
)

fig.show()

In [30]:
########### 5 ################
#Crear bins
df_tickets["Bin_N_Caracteres_Ticket"] = pd.qcut(
    df_tickets["N_Caracteres_Ticket"],
    q=3,
    labels=["Bajo", "Medio", "Alto"]
)

df_tickets["Bin_Tiempo_de_Resolución_Horas"] = pd.qcut(
    df_tickets["Tiempo_de_Resolución_Horas"],
    q=3,
    labels=["Bajo", "Medio", "Alto"]
)

df_tickets["Bin_Usuario_Antiguedad_Cuenta_Dias"] = pd.qcut(
    df_tickets["Usuario-Antiguedad_Cuenta_Dias"],
    q=3,
    labels=["Bajo", "Medio", "Alto"]
)

df_tickets[[
    "Bin_N_Caracteres_Ticket",
    "Bin_Tiempo_de_Resolución_Horas",
    "Bin_Usuario_Antiguedad_Cuenta_Dias"
]].head()

# Crear tablas cruzadas normalizadas por fila
ct_categoria = pd.crosstab(
    df_tickets["Categoría_Problema"],
    df_tickets["Nivel_Prioridad"],
    normalize="index"
)

ct_canal = pd.crosstab(
    df_tickets["Canal_Ticket"],
    df_tickets["Nivel_Prioridad"],
    normalize="index"
)

ct_tipo_cuenta = pd.crosstab(
    df_tickets["Usuario-Tipo_de_Cuenta"],
    df_tickets["Nivel_Prioridad"],
    normalize="index"
)

ct_caracteres = pd.crosstab(
    df_tickets["Bin_N_Caracteres_Ticket"],
    df_tickets["Nivel_Prioridad"],
    normalize="index"
)

ct_tiempo = pd.crosstab(
    df_tickets["Bin_Tiempo_de_Resolución_Horas"],
    df_tickets["Nivel_Prioridad"],
    normalize="index"
)

ct_antiguedad = pd.crosstab(
    df_tickets["Bin_Usuario_Antiguedad_Cuenta_Dias"],
    df_tickets["Nivel_Prioridad"],
    normalize="index"
)

#Graficar contingencias en subplots con imshow
tablas = [
    ct_categoria,
    ct_canal,
    ct_tipo_cuenta,
    ct_caracteres,
    ct_tiempo,
    ct_antiguedad
]

titulos = [
    "Categoría del Problema vs Nivel de Prioridad",
    "Canal del Ticket vs Nivel de Prioridad",
    "Tipo de Cuenta vs Nivel de Prioridad",
    "N° Caracteres Ticket vs Nivel de Prioridad",
    "Tiempo de Resolución vs Nivel de Prioridad",
    "Antigüedad Cuenta vs Nivel de Prioridad"
]

fig = make_subplots(
    rows=3,
    cols=2,
    subplot_titles=titulos
)

for i, tabla in enumerate(tablas):
    row = i // 2 + 1
    col = i % 2 + 1
    
    heatmap = go.Heatmap(
        z=tabla.values,
        x=tabla.columns,
        y=tabla.index,
        text=tabla.round(2).values,
        texttemplate="%{text}",
        showscale=False
    )
    
    fig.add_trace(heatmap, row=row, col=col)

fig.update_layout(
    title="Tablas de contingencia normalizadas por fila",
    height=1000
)

fig.show()

Ahora, responde las siguientes preguntas _(preguntas: 1.6 pts en total)_:

1. ¿Cuántos datos son? ¿Hay nulos? _(0.1 pts)_
2. ¿Qué columnas tenemos disponibles? ¿Qué tipo de dato es cada columna? - Tanto en tipo de dato (str, int, float, etc.) como también en su contenido (texto libre, categorías, entero, número decimal, fechas, índices) y su relación. Responde en una tabla con: Columna, Tipo de dato, Tipo de contenido y Descripción. _(0.2 pts)_
3. ¿Cómo se distribuyen las etiquetas? ¿Hay desbalance? Incluye porcentajes en la respuesta. _(0.2 pts)_
4. ¿Cómo afecta la distribución de clases a las respuestas dadas en la sección 1? ¿Sería necesario editar o corregir algo de estas? Si es así, edita las respuestas de la sección 1 y menciona qué edición hiciste aquí. _(0.1 pts)_
5. ¿Cómo se distribuyen las variables numéricas? ¿Con cuáles distribuciones las podríamos relacionar? ¿Cuál es la interpretación de estas? ¿Existen outliers? Explica por cada variable numérica. _(0.2 pts)_
6. ¿Existe alguna correlación entre las variables numéricas? _(0.1 pts)_
7. ¿Existe alguna relación interesante entre las categorías generadas y las categorías de la variable objetivo? _(0.2 pts)_
8. ¿Existe alguna relación entre el tiempo y la variable objetivo? ¿Sería recomendable utilizar esta variable? _(0.1 pts)_
9. Pensemos ahora por un momento en el modelo en modo inferencia/producción. ¿Existe alguna variable que no vaya a estar disponible en ese momento? Si es así, ¿cómo se llama este fenómeno? ¿Si no hubiésemos meditado esto, qué hubiese sucedido? _(0.2 pts)_
10. Para generar los bins de las variables numéricas se utilizó `qcut` (cuantiles) en lugar de `cut` (ancho fijo). ¿Qué habría pasado con el heatmap de contingencia de `Tiempo_de_Resolución_Horas` si hubieras usado `cut`? ¿Por qué `qcut` es preferible cuando la variable está muy sesgada? _(0.2 pts)_

> 1. El dataset contiene 2.000 registros y 11 columnas. Según el resultado de .info(), todas las columnas tienen 2.000 valores no nulos, por lo tanto no existen valores nulos en la base de datos. Esto es positivo para el análisis inicial, ya que no es necesario aplicar imputación o eliminación de registros por datos faltantes antes de entrenar el modelo.


> 2. 
| Columna | Tipo de dato | Tipo de contenido | Descripción |
|---|---|---|---|
| `Id_Ticket` | object | Identificador | Código único asociado a cada ticket. No debería usarse como variable predictiva, ya que solo identifica el registro. |
| `Asunto_Ticket` | object | Texto libre | Corresponde al asunto o título del ticket ingresado por el usuario. Puede aportar información textual útil para el modelo. |
| `Contenido_Ticket` | object | Texto libre | Contiene la descripción completa del problema reportado por el usuario. Es una de las variables más relevantes para construir el modelo. |
| `N_Caracteres_Ticket` | int64 | Entero / variable numérica | Indica la cantidad de caracteres del ticket. Es una variable derivada del texto y puede reflejar el nivel de detalle del reclamo. |
| `Canal_Ticket` | object | Categoría | Indica el canal por el cual ingresó el ticket, por ejemplo WhatsApp, Correo o Página Web. |
| `Categoría_Problema` | object | Categoría | Clasifica el tipo de problema reportado, como Cuenta, Técnica, Fraude, Cobros, Otro o Pregunta general. |
| `Usuario-Tipo_de_Cuenta` | object | Categoría | Indica el tipo de cuenta del usuario, por ejemplo Free, Premium o Business. |
| `Usuario-Antiguedad_Cuenta_Dias` | int64 | Entero / variable numérica | Representa la antigüedad de la cuenta del usuario en días. |
| `Fecha_Envio` | object | Fecha | Corresponde a la fecha de envío del ticket. Aunque aparece como `object`, puede convertirse a formato fecha con `pd.to_datetime`. |
| `Tiempo_de_Resolución_Horas` | int64 | Entero / variable numérica | Indica el tiempo que tomó resolver el ticket en horas. Debe analizarse con cuidado porque probablemente no está disponible al momento de predecir. |
| `Nivel_Prioridad` | object | Variable objetivo categórica | Es la variable que se busca predecir. Sus categorías son Baja, Media, Alta y Critica. |

> 3. La variable objetivo es Nivel_Prioridad. A partir del histograma, se observa que las clases no están distribuidas de forma completamente uniforme. La distribución aproximada para el nivel de prioridad Media tiene una cantidad a proximada de 790, que equivale a 39,5%. Para el nivel de prioridad baja tiene una cantidad aproximada de 720, que equivale a 36%. Para el nivel de prioridad Alta tiene una cantidad aproximada de 350, que equivale a 17,5%. Y para el nivel de prioridad Crítica tiene una cantidad aproximada de 140, que equivale a 7%. \\
Sí existe desbalance de clases, ya que las clases Media y Baja concentran la mayor parte de los tickets, mientras que Critica representa una proporción mucho menor. Esto es relevante porque un modelo podría obtener buen accuracy clasificando correctamente las clases más frecuentes, pero fallar en la detección de tickets críticos.

> 4. La distribución de clases refuerza lo planteado en la sección 1, no conviene evaluar el modelo solo con accuracy, porque las clases están desbalanceadas. Si el modelo aprende principalmente a clasificar tickets como Baja o Media, podría lograr un buen desempeño global, pero fallar en las clases más relevantes para el negocio, como Alta y Critica.Por esto, es importante mantener como métricas principales el recall de las clases Alta y Critica, junto con el macro F1-score, ya que esta métrica da el mismo peso a todas las clases y permite evaluar mejor el desempeño en clases minoritarias.

> 5. Las variables numéricas disponibles son N_Caracteres_Ticket, Usuario-Antiguedad_Cuenta_Dias y Tiempo_de_Resolución_Horas.
> - N_Caracteres_Ticket: Esta variable tiene una media de aproximadamente 718 caracteres, una mediana de 697 y un máximo de 1.831. Su distribución es unimodal y presenta una leve asimetría hacia la derecha, ya que existen tickets mucho más largos que el promedio. Podría relacionarse con una distribución normal levemente sesgada o una distribución log-normal. Existen posibles outliers en tickets con muchos caracteres.
> - Usuario-Antiguedad_Cuenta_Dias: Esta variable presenta una media de aproximadamente 122 días, una mediana de 78 días y un máximo de 1.200 días. La distribución está fuertemente sesgada hacia la derecha, la mayoría de los usuarios tiene cuentas relativamente nuevas, pero existen algunos usuarios con cuentas muy antiguas. Esta variable podría relacionarse con una distribución exponencial o log-normal. Sí existen outliers, especialmente en cuentas con antigüedad muy alta.
> - Tiempo_de_Resolución_Horas: Esta variable tiene una media de aproximadamente 18,7 horas, una mediana de 13 horas y un máximo de 183 horas. También presenta una distribución sesgada hacia la derecha, ya que la mayoría de los tickets se resuelve en pocas horas, pero algunos tienen tiempos de resolución muy altos. Esta variable podría relacionarse con una distribución exponencial, gamma o log-normal. Existen outliers en tickets con tiempos de resolución muy elevados.

> 6. La matriz de correlación muestra que las relaciones entre variables numéricas son débiles. Los valores observados son aproximadamente: N_Caracteres_Ticket y Tiempo_de_Resolución_Horas, 0,19. N_Caracteres_Ticket y Usuario-Antiguedad_Cuenta_Dias, 0,07. Usuario-Antiguedad_Cuenta_Dias y Tiempo_de_Resolución_Horas, 0,02. La correlación más alta es entre N_Caracteres_Ticket y Tiempo_de_Resolución_Horas, con un valor cercano a 0,19, pero sigue siendo una correlación baja. Esto indica que tickets más largos podrían tender a demorar un poco más en resolverse, pero la relación no es fuerte.

> 7. Sí, se observan varias relaciones interesantes entre las variables categóricas y Nivel_Prioridad. En Categoría_Problema, los tickets de Fraude presentan una mayor proporción de prioridades altas y críticas. Aproximadamente un 37% de los tickets de fraude son Alta y un 24% son Critica. Esto tiene sentido desde el punto de vista del negocio, porque problemas asociados a fraude suelen requerir atención urgente.\\
En cambio, la categoría Pregunta general se concentra principalmente en prioridades bajas o medias. En el heatmap se observa que cerca del 70% de estos tickets son Baja y alrededor del 30% son Media, sin presencia relevante de Alta o Critica.\\
También se observa que los tickets de usuarios Business tienen mayor proporción de prioridades altas y críticas. Aproximadamente un 41% corresponde a Alta y un 22% a Critica. Esto sugiere que el tipo de cuenta puede ser una variable relevante para el modelo.\\
Respecto al canal, los tickets ingresados por Correo muestran una mayor proporción de casos Alta y Critica en comparación con WhatsApp. Esto podría indicar que ciertos canales concentran problemas más complejos o urgentes.\\
Finalmente, al analizar los bins de N_Caracteres_Ticket, se observa que los tickets con mayor cantidad de caracteres tienen una mayor proporción de prioridad Alta y Critica, mientras que los tickets más cortos se concentran principalmente en prioridad Baja.

> 8. Sí, existe cierta relación entre Tiempo_de_Resolución_Horas y Nivel_Prioridad. En el heatmap se observa que los tickets con mayor tiempo de resolución tienen una mayor proporción de prioridades Alta y Critica que los tickets con bajo tiempo de resolución.
Por ejemplo, en el grupo de tiempo bajo, la prioridad Critica representa aproximadamente un 4%, mientras que en el grupo de tiempo alto representa cerca de un 9%. También aumenta la proporción de tickets Alta.
Sin embargo, no sería recomendable usar Tiempo_de_Resolución_Horas como variable predictiva para el modelo de inferencia. Esto se debe a que el tiempo de resolución solo se conoce después de que el ticket fue atendido y cerrado. Como el objetivo del modelo es priorizar tickets al momento de recibirlos, esta información no estaría disponible en producción. Por lo tanto, aunque la variable tenga relación con la prioridad, usarla podría generar un problema de fuga de información.

> 9. Sí. La variable más problemática es Tiempo_de_Resolución_Horas, porque representa una información posterior al ingreso del ticket. Al momento en que un usuario envía un nuevo ticket, todavía no se sabe cuánto tiempo demorará en resolverse.
Este fenómeno se llama data leakage o, más específicamente, target leakage. Ocurre cuando el modelo utiliza información que no estaría disponible al momento real de hacer la predicción, o información que está directa o indirectamente influenciada por la variable objetivo.
Si no se considera este problema, el modelo podría obtener muy buenos resultados durante la evaluación, pero esos resultados serían engañosos. En producción, el modelo no tendría acceso a esa variable y su desempeño real sería mucho peor. Por esta razón, Tiempo_de_Resolución_Horas debería excluirse del entrenamiento del modelo predictivo. También se debe excluir Id_Ticket, ya que es solo un identificador y no contiene información generalizable para predecir la prioridad.

> 10. Si se hubiera usado cut, los bins se habrían generado con intervalos de ancho fijo. Esto puede ser problemático cuando la variable está muy sesgada, como ocurre con Tiempo_de_Resolución_Horas. En este caso, la mayoría de los tickets se concentra en tiempos bajos de resolución, mientras que existen pocos tickets con tiempos muy altos. Si se usara cut, probablemente el primer intervalo concentraría una gran cantidad de observaciones y los intervalos superiores quedarían con muy pocos datos. Esto haría que el heatmap de contingencia fuera menos estable y más difícil de interpretar, porque algunas filas tendrían muy pocos casos. \\
En cambio, qcut genera bins según cuantiles, por lo que intenta dejar una cantidad similar de observaciones en cada grupo: Bajo, Medio y Alto. Esto es preferible en variables sesgadas porque permite comparar proporciones entre grupos más equilibrados.

---

## Subsección 2 - Embeddings

_(Total: 1.4 pts — código 0.6 + preguntas 0.8)_

En esta sección trabajarás con embeddings sobre los Asuntos + Descripciones de tickets precalculados con el modelo `gemini-embeddings-001`. Ahora, nuevamente utilizando las herramientas de lectura de parquet _(código: 0.6 pts en total)_:

1. Lee el archivo que contiene los embeddings (`data/embeddings.parquet`). _(0.1 pts)_
2. Reduce la dimensionalidad de los embeddings con umap-learn a dos dimensiones y agrega a `df_tickets` las nuevas dimensiones. Asegura tu proyección utilizando semillas, así luego, al momento de la revisión, se podrán replicar los resultados. _(0.3 pts)_
3. Grafica un scatter plot con las proyecciones y pinta por `Nivel_Prioridad` (utiliza el código dado).
4. Grafica un nuevo scatter plot con las proyecciones y pinta por `Categoría_Problema`. _(0.2 pts)_

In [31]:
import pandas as pd
import numpy as np
import plotly.express as px
from pathlib import Path

from umap import UMAP



In [49]:
#Leer el archivo embeddings.parquet
ruta_embeddings = Path("data/embeddings.parquet")

df_embeddings = pd.read_parquet(ruta_embeddings)

df_embeddings.head()

,embedding_dim_1,embedding_dim_2,embedding_dim_3,embedding_dim_4,embedding_dim_5,embedding_dim_6,embedding_dim_7,embedding_dim_8,embedding_dim_9,embedding_dim_10,...,embedding_dim_1015,embedding_dim_1016,embedding_dim_1017,embedding_dim_1018,embedding_dim_1019,embedding_dim_1020,embedding_dim_1021,embedding_dim_1022,embedding_dim_1023,embedding_dim_1024
Id_Ticket,,,,,,,,,,,,,,,,,,,,,
TCK-19267,-0.011585,-0.008728,0.012878,-0.050069,0.006369,0.008990,-0.008538,0.010781,-0.021699,0.007514,...,-0.011539,-0.016249,0.004422,-0.022274,0.004511,0.001428,-0.007812,-0.020485,-0.001888,-0.000408
TCK-19268,0.014494,0.004070,0.011298,-0.056227,0.009839,0.014777,0.005187,0.001136,-0.000801,-0.006830,...,-0.019967,-0.008424,-0.002084,-0.002761,-0.004148,-0.007131,-0.012717,-0.019045,-0.015172,0.004308
TCK-19269,-0.010936,0.020796,0.011680,-0.080212,0.017745,0.020321,0.004509,0.000413,-0.009146,0.008450,...,-0.021763,-0.007094,-0.011453,-0.001084,-0.005820,-0.015085,0.003200,-0.019789,-0.018502,-0.001085
TCK-19270,0.003248,0.005420,0.003422,-0.065296,0.026466,0.018159,-0.011115,0.004692,0.001486,-0.006940,...,-0.019830,-0.007490,-0.013525,0.000086,-0.011983,0.007148,-0.002536,-0.016200,-0.003927,-0.004028
TCK-19271,-0.010437,0.013634,0.009859,-0.066499,0.006272,0.006100,-0.010604,0.009686,-0.032323,-0.003181,...,-0.006401,-0.001971,0.002344,-0.010598,-0.011034,-0.012039,-0.000857,-0.009420,-0.006597,0.012463


In [50]:
df_embeddings.info()
df_embeddings.columns

<class 'pandas.core.frame.DataFrame'>
Index: 2000 entries, TCK-19267 to TCK-21266
Columns: 1024 entries, embedding_dim_1 to embedding_dim_1024
dtypes: float64(1024)
memory usage: 15.6+ MB


Index(['embedding_dim_1', 'embedding_dim_2', 'embedding_dim_3',
       'embedding_dim_4', 'embedding_dim_5', 'embedding_dim_6',
       'embedding_dim_7', 'embedding_dim_8', 'embedding_dim_9',
       'embedding_dim_10',
       ...
       'embedding_dim_1015', 'embedding_dim_1016', 'embedding_dim_1017',
       'embedding_dim_1018', 'embedding_dim_1019', 'embedding_dim_1020',
       'embedding_dim_1021', 'embedding_dim_1022', 'embedding_dim_1023',
       'embedding_dim_1024'],
      dtype='object', length=1024)

In [51]:
#Alinear embeddings con df_tickets
# Revisar tamaños
print("Tickets:", df_tickets.shape)
print("Embeddings:", df_embeddings.shape)

# Revisar primeros índices de embeddings
df_embeddings.index[:5]

Tickets: (2000, 14)
Embeddings: (2000, 1024)


Index(['TCK-19267', 'TCK-19268', 'TCK-19269', 'TCK-19270', 'TCK-19271'], dtype='object', name='Id_Ticket')

In [52]:
# Alinear embeddings con el orden de df_tickets usando Id_Ticket
df_embeddings_ordenado = df_embeddings.loc[df_tickets["Id_Ticket"]]

# Verificar que quedaron en el mismo orden
print(df_embeddings_ordenado.shape)
print(df_tickets["Id_Ticket"].head())
print(df_embeddings_ordenado.index[:5])

(2000, 1024)
0    TCK-19267
1    TCK-19268
2    TCK-19269
3    TCK-19270
4    TCK-19271
Name: Id_Ticket, dtype: object
Index(['TCK-19267', 'TCK-19268', 'TCK-19269', 'TCK-19270', 'TCK-19271'], dtype='object', name='Id_Ticket')


In [53]:
#Crear matriz de embeddings

X_embeddings = df_embeddings_ordenado.values

print("Shape matriz embeddings:", X_embeddings.shape)

Shape matriz embeddings: (2000, 1024)


In [54]:
#Aplicamos UMAP
umap_model = UMAP(
    n_components=2,
    random_state=42,
    n_neighbors=15,
    min_dist=0.1
)

proyecciones = umap_model.fit_transform(X_embeddings)

proyecciones[:5]

c:\Users\Acer\OneDrive\Escritorio\MDS7202\.venv\lib\site-packages\umap\umap_.py:1952: UserWarning: n_jobs value 1 overridden to 1 by setting random_state. Use no seed for parallelism.
  warn(


array([[13.969598 ,  3.3063364],
       [12.176239 ,  1.4680829],
       [-1.8042221,  2.9645345],
       [11.579033 ,  4.6340547],
       [ 8.634287 ,  5.318221 ]], dtype=float32)

In [55]:
# Nuevas dimensiones a df_tickets
df_tickets["x"] = proyecciones[:, 0]
df_tickets["y"] = proyecciones[:, 1]

df_tickets[["Id_Ticket", "x", "y", "Nivel_Prioridad", "Categoría_Problema"]].head()

,Id_Ticket,x,y,Nivel_Prioridad,Categoría_Problema
0,TCK-19267,13.969598,3.306336,Baja,Cuenta
1,TCK-19268,12.176239,1.468083,Media,Otro
2,TCK-19269,-1.804222,2.964535,Baja,Fraude
3,TCK-19270,11.579033,4.634055,Media,Cuenta
4,TCK-19271,8.634287,5.318221,Media,Otro


In [57]:
#Scatter plot por Nivel_Prioridad
import textwrap


def wrap_text(text, width=40):
    return "<br>".join(textwrap.wrap(text, width=width))


df_tickets["Contenido_Ticket_Wrapped"] = df_tickets["Contenido_Ticket"].apply(lambda x: wrap_text(x, 80))

fig_proj = px.scatter(
    df_tickets,
    x="x",
    y="y",
    color="Nivel_Prioridad",
    title="Proyecciones de los embeddings respecto a clase",
    hover_data=[
        "Canal_Ticket",
        "N_Caracteres_Ticket",
        "Categoría_Problema",
        "Usuario-Tipo_de_Cuenta",
        "Usuario-Antiguedad_Cuenta_Dias",
        "Contenido_Ticket_Wrapped",
    ],
)
fig_proj.show()

In [58]:
# Scatter plot por Categoría_Problema
fig_proj_categoria = px.scatter(
    df_tickets,
    x="x",
    y="y",
    color="Categoría_Problema",
    title="Proyecciones de los embeddings respecto a Categoría del Problema",
    hover_data=[
        "Nivel_Prioridad",
        "Canal_Ticket",
        "N_Caracteres_Ticket",
        "Usuario-Tipo_de_Cuenta",
        "Usuario-Antiguedad_Cuenta_Dias",
        "Contenido_Ticket_Wrapped",
    ],
)

fig_proj_categoria.show()

Ahora, responde las siguientes preguntas _(preguntas: 0.8 pts en total)_:

1. ¿Qué es un embedding? ¿Qué representan? A grandes rasgos, ¿cómo se calculan? Hint: consulta sobre `gemini-embedding-001`. _(0.2 pts)_
2. ¿Por qué los embeddings tienen el potencial de ser mejores para un modelo predictivo que analizar las palabras individuales? _(0.1 pts)_
3. ¿Cuál podría ser una desventaja de utilizar embeddings en vez de utilizar directamente palabras? _(0.1 pts)_
4. ¿Existen otros tipos de embeddings? ¿Qué podrían representar? _(0.1 pts)_
5. ¿Qué patrón interesante puede mostrar el gráfico de proyecciones pintadas por `Nivel_Prioridad`? _(0.1 pts)_
6. ¿Qué patrón interesante puede mostrar el gráfico de proyecciones pintadas por `Categoría_Problema`? _(0.1 pts)_
7. Las proyecciones de UMAP (`x`, `y`) quedaron agregadas a `df_tickets`. ¿Sería buena idea usar estas 2 dimensiones como features de entrada del modelo predictivo, en lugar de los embeddings originales? Justifica pensando en la información que se pierde al reducir a 2D. _(0.1 pts)_

> 1. Un embedding es una representación numérica de un dato, en este caso texto, mediante un vector de muchas dimensiones. En vez de trabajar directamente con palabras, el modelo transforma el contenido textual del ticket en una secuencia de números que intenta capturar su significado, contexto y similitud semántica. Para este proyecto, los embeddings fueron calculados con el modelo gemini-embedding-001, que recibe texto como entrada y entrega embeddings de texto como salida. Según la documentación de Google AI, este modelo soporta dimensiones de salida flexibles entre 128 y 3072, con tamaños recomendados de 768, 1536 y 3072 dimensiones. En nuestro caso, el archivo embeddings.parquet contiene 1024 dimensiones para cada ticket, desde embedding_dim_1 hasta embedding_dim_1024.

> 2. Los embeddings pueden ser mejores que analizar palabras individuales porque no solo consideran la presencia de términos aislados, sino también el significado general del texto. Por ejemplo, frases como “no puedo pagar”, “la compra fue rechazada” o “mi tarjeta no funciona” pueden describir problemas parecidos aunque usen palabras distintas.\\
Un enfoque basado solo en palabras individuales puede perder estas relaciones semánticas. En cambio, los embeddings permiten que textos similares queden más cercanos entre sí, incluso si no comparten exactamente las mismas palabras. Esto puede ayudar a que el modelo predictivo detecte patrones más robustos en los tickets.

> 3. Una desventaja es que los embeddings son menos interpretables. Cuando se trabaja con palabras directamente, es más fácil explicar que cierta palabra o frase influyó en la predicción. En cambio, un embedding está compuesto por muchas dimensiones numéricas, por lo que no es evidente qué representa cada dimensión individual. Además, usar embeddings puede aumentar la complejidad computacional del modelo. En este caso, cada ticket tiene 1024 variables provenientes del embedding, lo que puede hacer que el entrenamiento sea más costoso y más difícil de interpretar.

> 4. Sí, existen distintos tipos de embeddings. En este proyecto se utilizan embeddings de texto, que representan el significado del asunto y contenido del ticket. Sin embargo, también existen embeddings para imágenes, audio, video, usuarios, productos o documentos.
Por ejemplo, un embedding de imagen podría representar características visuales de una fotografía, un embedding de audio podría representar patrones sonoros, y un embedding de usuario podría representar su comportamiento o preferencias.

> 5. En el gráfico coloreado por Nivel_Prioridad, se observa que las clases no quedan completamente separadas. Las prioridades Baja y Media aparecen bastante mezcladas en varias zonas del plano, lo que sugiere que, solo a partir de la proyección 2D, no existe una frontera clara entre estas clases. Sin embargo, también se observan algunas zonas donde aparecen más concentrados tickets Alta y Critica, especialmente en ciertos grupos específicos. Esto sugiere que los embeddings sí capturan información útil relacionada con la prioridad, pero la separación no es perfecta. Por lo tanto, el modelo probablemente necesitará combinar los embeddings con otras variables del ticket para mejorar la predicción.

> 6. En el gráfico coloreado por Categoría_Problema, se observa una separación más clara que en el gráfico por prioridad. Algunas categorías tienden a formar grupos más definidos en el espacio de embeddings. Esto tiene sentido, porque los embeddings se construyen a partir del contenido textual del ticket, y el texto suele estar directamente relacionado con el tipo de problema reportado. Por ejemplo, tickets asociados a temas como cobros, fraude, cuenta, técnica o pregunta general tienden a ocupar zonas distintas del gráfico. Esto indica que los embeddings capturan bastante bien diferencias semánticas entre tipos de reclamos. Entonces, la información textual parece ser especialmente útil para identificar la categoría del problema y también podría aportar indirectamente a la predicción de la prioridad.

> 7. No sería recomendable usar únicamente las dos dimensiones x e y generadas por UMAP como variables de entrada principales del modelo. Estas dimensiones son útiles para visualizar los embeddings en un plano, pero al reducir los vectores originales desde 1024 dimensiones a solo 2 dimensiones se pierde una gran cantidad de información. UMAP intenta conservar relaciones de cercanía entre puntos, pero su objetivo principal en este caso es la visualización, no necesariamente maximizar el desempeño predictivo. Por eso, usar solo x e y podría simplificar demasiado el problema y hacer que el modelo pierda detalles semánticos importantes del texto.